addons/maintainence-agent/maintenance_mode_controller.go

```diff
@@ -54,66 +54,75 @@ func (r *MaintenanceModeReconciler) Reconcile(ctx context.Context, req ctrl.Requ
-				return ctrl.Result{}, err
-			}
-		}
-	}
-
-	cephClusters, err := fetchAllCephClusters(ctx, r.SpokeClient)
+	cephClusters, err := utils.FetchAllCephClusters(ctx, r.SpokeClient)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
 	if cephClusters == nil || len(cephClusters.Items) == 0 {
 		klog.Info("no CephClusters available on the cluster")
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
 	}
 
-	actionableCephClusters := filterCephClustersByStorageId(cephClusters, mmode.Spec.TargetID, mmode.Spec.StorageProvisioner)
+	actionableCephCluster := filterCephClustersByStorageIdOrReplicationId(cephClusters, mmode.Spec.TargetID, mmode.Spec.StorageProvisioner)
 
-	if len(actionableCephClusters) == 0 {
-		klog.Info("no CephClusters present with required parameters for maintenance", "StorageId", mmode.Spec.TargetID, "Provisioner", mmode.Spec.StorageProvisioner)
-		return ctrl.Result{}, nil
+	if actionableCephCluster == nil {
+		klog.Infof("no CephCluster present with required parameters for maintenance. StorageId/ReplicationId=%s , Provisioner=%s ", mmode.Spec.TargetID, mmode.Spec.StorageProvisioner)
+		// Requeueing the request as the cephcluster can be potentially found if it is labeled later.
+		return ctrl.Result{Requeue: true}, nil
 	}
 
 	if !mmode.GetDeletionTimestamp().IsZero() || len(mmode.Spec.Modes) == 0 {
-		result, err := r.startMaintenanceActions(ctx, mmode, actionableCephClusters, false)
+		if !utils.ContainsString(mmode.GetFinalizers(), MaintenanceModeFinalizer) {
+			return ctrl.Result{}, nil
+		}
+		result, err := r.startMaintenanceActions(ctx, mmode, actionableCephCluster, false)
 		if err != nil {
 			return result, err
 		}
 		mmode.Finalizers = utils.RemoveString(mmode.Finalizers, MaintenanceModeFinalizer)
-
-		if err := r.SpokeClient.Update(ctx, &mmode); err != nil {
-			klog.Error(err, "failed to remove finalizer from MaintenanceMode ", err)
+		err = r.SpokeClient.Update(ctx, &mmode)
+		if err != nil {
+			klog.Error("failed to remove finalizer from MaintenanceMode ")
 			return ctrl.Result{}, err
 		}
-		klog.Info("MaintenanceMode deleted, skipping reconciliation")
+		klog.Info("MaintenanceMode disabled")
 		return ctrl.Result{}, nil
-	} else {
-		_, err := r.startMaintenanceActions(ctx, mmode, actionableCephClusters, true)
-		if err != nil {
+	}
+
+	if !utils.ContainsString(mmode.GetFinalizers(), MaintenanceModeFinalizer) {
+		klog.Infof("finalizer not found on MaintenanceMode. adding Finalizer=%s", MaintenanceModeFinalizer)
+		mmode.Finalizers = append(mmode.Finalizers, MaintenanceModeFinalizer)
+		if err := r.SpokeClient.Update(ctx, &mmode); err != nil {
+			klog.Errorf("failed to add finalizer to MaintenanceMode=%s. err=%v ", mmode.Name, err)
 			return ctrl.Result{}, err
 		}
 	}
 
-	return ctrl.Result{}, nil
+	if mmode.Status.State == ramenv1alpha1.MModeStateCompleted && mmode.Status.ObservedGeneration == mmode.Generation {
+		statusErr := r.SpokeClient.Status().Update(ctx, &mmode)
```

Comment:
Why update status here? We did not process anything for the request till now, and have determined that status as per the API server is completed for the latest generation, so there really is no status to update.